src/libOpenImageIO/imagebuf.cpp

```
@@ -685,38 +737,47 @@ ImageBuf::operator=(ImageBuf&& src)
 
 
 char*
-ImageBufImpl::new_pixels(size_t size, const void* data)
-{
-    if (m_allocated_size)
-        free_pixels();
-    try {
-        m_pixels.reset(size ? new char[size] : nullptr);
-        // Set bufspan to the allocated memory
-        m_bufspan = { reinterpret_cast<std::byte*>(m_pixels.get()), size };
-    } catch (const std::exception& e) {
-        // Could not allocate enough memory. So don't allocate anything,
-        // consider this an uninitialized ImageBuf, issue an error, and hope
-        // it's handled well downstream.
-        m_pixels.reset();
-        m_bufspan = make_span<std::byte>(nullptr, 0);
-        OIIO::debugfmt("ImageBuf unable to allocate {} bytes ({})\n", size,
-                       e.what());
-        error("ImageBuf unable to allocate {} bytes ({})\n", size, e.what());
-        size      = 0;
-        m_bufspan = {};
+ImageBufImpl::new_pixels(ImageBuf::IBStorage storage, size_t size,
+                         const void* data)
+{
+    m_storage = storage;
+    if (storage == ImageBuf::LOCALBUFFER && !m_spec.deep)
+        OIIO_ASSERT(size == m_spec.image_bytes());
+    size = (storage == ImageBuf::LOCALBUFFER && !m_spec.deep)
+               ? m_spec.image_bytes()
+               : 0;
+    if (m_allocated_size != size) {
+        if (m_allocated_size)
+            free_pixels();
+        try {
+            m_pixels.reset(size ? new char[size] : nullptr);
+            // Set m_bufspan to the allocated memory
+            set_bufspan(m_pixels.get());
+        } catch (const std::exception& e) {
```

[consider] If this is truly meant to only catch out of memory exceptions this should explicitly catch `std::bad_alloc` to still allow other exceptions to gracefully make their way up the stack

---

We try not to let exceptions propagate, since nothing called in OIIO's public APIs is supposed to throw. And I'm not sure what else it would be here? But I'm being over-inclusive.